src/librustc_mir/interpret/operand.rs

```diff
@@ -234,6 +234,12 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         {
             Some(ptr) => ptr,
             None => {
+                if let Scalar::Ptr(ptr) = mplace.ptr {
+                    // If the ZST pointer has provenance, we may be reading from a static.
+                    // In order to ensure that `static FOO: Type = FOO;` causes a cycle, we need to
+                    // trigger a read here.
```

Comment:
Why do this here and not centrally in `Memory::check_ptr_access`?

Reply:
Because that function doesn't actually do any reads so far in order to prevent other kind of cycles. I'm not sure whether any of the call sites need this.

Reply:
Moving the check worked.

Reply:
So, moving this has started to cause problems, because validation goes through the `check_ptr_access` without actually doing a read (https://github.com/rust-lang/rust/issues/71612#issuecomment-622390064) and because const prop now creates references to foreign statics that later code ICEs on https://github.com/rust-lang/rust/issues/71709#issuecomment-622121249